accepted/00000-modular-repos-with-clm.md

```diff
@@ -0,0 +1,179 @@
+```
+
+The goal is to handle modular repositories in a way that the clients can be still managed through the Uyuni UI.
+
+# Detailed design
+[design]: #detailed-design
+
+## Overview
+
+Instead of implementing the modularity logic inside the core of Uyuni, this document describes a design to simulate the modularity behavior by building regular repositories out of modular repositories by filtering out packages from unselected streams.
+The resulting repository is a "flat" repository with no modularity, so it can be managed with Uyuni's existing package operations logic.
+
+Conceptually, a module can be in one of the following three states at any time:
+ - Enabled
+ - Disabled
+ - Neither enabled nor disabled
+
+When a module is neither enabled nor disabled, all the alternative streams and contexts of a stream are available to the package manager. This state confuses Uyuni when comparing the package versions.
+
+The key to building a regular repository with no conflicting packages is to simulate the modules in a way that every single module is either "enabled" or "disabled".
+
+Therefore, the high-level algorithm is proposed as:
+ 1. Let the users pick desired streams so that every module is either "enabled" or "disabled"
+ 2. Compile a blacklist of RPM artifacts from the "disabled" modules
+ 3. Generate a "regular" repository, filtering out the blacklisted RPMs
+
+The design utilizes content lifecycle management filters and repository building capabilities to implement the algorithm above.
+
+## Components
+
+This section summarizes the different components involved in the design.
+
+### Backend
+The Java backend is responsible for filtering out unselected module streams and building the repository with the desired packages through CLM.
+
+### API
+A Python API parses and interprets the module metadata, resolves modular dependencies and translates specified modules and streams to an actual package list.
+The API utilizes the Python port of [libmodulemd](https://github.com/fedora-modularity/libmodulemd), which is used for structural parsing of `modules.yaml` files.
+
+An example of such implementation is provided in [attachments/00000-modular-repos-api.py](attachments/00000-modular-repos-api.py).
+
+### UI
+The UI consists of a new type of CLM filter called "module". The filter provides inputs for a module name and a stream name.
+Each filter selects a stream for a single module. The stream input can be left empty to select the default stream.
+Since a CLM filter is not related to any channels, it is not possible to display a list of modules/streams on the filter form and the inputs can only be simple text inputs.
+The module information for a modular repository is displayed under an additional "Modules" tab on the "Channel Details" page.
+A modular repository is indicated with an icon next to the name of the repository in the CLM sources list. A click on the icon navigates to the "Modules" tab of the channel details.
```

Comment:
Assuming we won't have any feedback in the filter form (simple text inputs) this would be the only way to provide information about the modules. So I'd say it's mandatory.

Reply:
Agreed. This highly depends on how we implement the filter.